### PR TITLE
AI: architecture + logic

### PR DESCRIPTION
**Pull request by deividvd**
# Description

Added AI project

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

The tests are games simulations, represented using only board

- [x] LevelZeroSpec
- [x] LevelOneSpec
- [x] LevelTwoSpec

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: ai/src/main/scala/it/scalachess/ai/level/LevelOne.scala [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+import it.scalachess.core.Color
+import it.scalachess.core.board.Board
+import it.scalachess.core.logic.moves.FullMove
+import it.scalachess.core.logic.moves.generators.MoveGenerator
+import it.scalachess.core.pieces.{ Bishop, King, Knight, Pawn, PieceType, Queen, Rook }
+
+/**
+ * The level one AI plays the move which capture the more important piece.
+ */
+case class LevelOne() extends Level {
```
Comment (CastFX) [inline ai/src/main/scala/it/scalachess/ai/level/LevelOne.scala]: final case class

File: ai/src/main/scala/it/scalachess/ai/level/LevelTwo.scala [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+        fullMove -> minimax(board(fullMove.validMove.boardChanges),
+                            history :+ fullMove,
+                            depth - 1,
+                            aiPlayer.other,
+                            aiPlayer)
+      })
+      .toMap
+  }
+
+  def minimax(board: Board,
```
Comment (CastFX) [inline ai/src/main/scala/it/scalachess/ai/level/LevelTwo.scala:56]: needs documentation

File: ai/src/main/scala/it/scalachess/ai/level/LevelTwo.scala [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+import it.scalachess.core.board.Board
+import it.scalachess.core.logic.moves.FullMove
+import it.scalachess.core.logic.moves.generators.MoveGenerator
+
+/**
+ * The level two AI plays the best move following two principles:
+ * [1] it cares about value based on piece importance (it uses the same evaluation of level one);
+ * [2] it analyse all the possible next moves, relying on the minimax algorithm.
+ */
+case class LevelTwo() extends Level {
```
Comment (CastFX) [inline ai/src/main/scala/it/scalachess/ai/level/LevelTwo.scala]: final case class

File: ai/src/main/scala/it/scalachess/ai/level/LevelZero.scala [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+import it.scalachess.core.board.Board
+import it.scalachess.core.logic.moves.FullMove
+import it.scalachess.core.logic.moves.generators.MoveGenerator
+
+import scala.util.Random
+
+/**
+ * The level zero AI plays random moves.
+ */
+case class LevelZero() extends Level {
```
Comment (CastFX) [inline ai/src/main/scala/it/scalachess/ai/level/LevelZero.scala]: final case class

File: ai/src/main/scala/it/scalachess/ai/level/Level.scala [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+trait Level extends ((Board, Color, Seq[FullMove]) => FullMove) {
+
+  override def apply(board: Board, aiPlayer: Color, history: Seq[FullMove]): FullMove
+
+  protected def moveWithMaxEvaluation(movesEvaluated: Map[FullMove, Double]): FullMove = {
+    assert(movesEvaluated.nonEmpty, aiPlayerInCheckmateFailMsg)
+    movesEvaluated.find(_._2 == movesEvaluated.values.max).get._1
+  }
+
+  protected val othePlayerInCheckmate      = "The other player is in checkmate: wrong usage of AI"
```
Comment (ZeroMaicol) [inline ai/src/main/scala/it/scalachess/ai/level/Level.scala]: If used, correct othe in other, else delete

**Review by CastFX — CHANGES_REQUESTED** [review]

**Review by ZeroMaicol — APPROVED** [review]

**Review by CastFX — APPROVED** [review]

**Pull request merged**